### PR TITLE
feat: sdk batch rpc calls

### PR DESCRIPTION
**Pull request by abretonc7s**
## **Description**

Allow dapp developers to create sequences of RPC method calls that are sent simultaneously and managed by the MM mobile app using the existing RPC queue.

Chained RPC Calls:

Implement the ability for the SDK to handle sequences of RPC method calls.
Address known issues in the RPC queue, especially when one of the methods fails.

## **Related issues**

[Fixes: sdk 391
](https://github.com/MetaMask/metamask-sdk/issues/391)

## **Manual testing steps**

1. Use a dapp using metamask/sdk and call the following:
```
    const selectedAddress = provider?.selectedAddress;

    const rpcs: ChainRPC[] = [{
      method: "personal_sign",
      params: ["something to sign 1", selectedAddress],
    },
    {
      method: "personal_sign",
      params: ["hello world", selectedAddress],
    },
    {
      method: "personal_sign",
      params: ["Another one #3", selectedAddress],
    }];

    try {
      const response = await provider?.request({method: 'metamask_batch', params: rpcs}) as any[];
      ...
```

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/core/SDKConnect/BatchRPCManager.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+  remove(id: string) {
+    delete this.rpcChain[id];
+  }
+
+  get() {
+    return this.rpcChain;
+  }
+
+  getId(id: string) {
```
Comment (omridan159) [inline app/core/SDKConnect/BatchRPCManager.ts:60]:
```suggestion
  getRpcStateById(id: string) {
```
WDYT?

File: app/core/SDKConnect/BatchRPCManager.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
+interface RPCMethod {
+  id: string;
+  method: string;
+  params: any;
+  jsonrpc: string;
+  response?: any;
+}
```
Comment (omridan159) [inline app/core/SDKConnect/BatchRPCManager.ts:7]:
```suggestion
interface RPCMethod<TParams = any, TResponse = any> {  id: string;
  method: string;
  params: TParams;
  jsonrpc: string;
  response?: TResponse;
}
```
WDYT?

File: app/core/SDKConnect/Connection.ts [hunk 3]
```diff
@@ -452,6 +473,38 @@ export class Connection extends EventEmitter2 {
           Logger.log(`metamask_connectSign`, message.params);
         }
 
+        if (message.method === 'metamask_batch') {
+          DevLogger.log(`metamask_batch`, JSON.stringify(message, null, 2));
+          if (
+            !(
+              message.params &&
+              Array.isArray(message?.params) &&
```
Comment (omridan159) [inline app/core/SDKConnect/Connection.ts:481]:
```suggestion
              Array.isArray(message.params) &&
```

Comment (omridan159) [reply app/core/SDKConnect/Connection.ts:481]: Redundant Null Check

File: app/core/SDKConnect/Connection.ts [hunk 4]
```diff
@@ -452,6 +473,38 @@ export class Connection extends EventEmitter2 {
           Logger.log(`metamask_connectSign`, message.params);
         }
 
+        if (message.method === 'metamask_batch') {
```
Comment (omridan159) [inline app/core/SDKConnect/Connection.ts:476]: It might be a good idea to put all the method types in a const

**Review by omridan159 — CHANGES_REQUESTED** [review]
.

**Review by omridan159 — DISMISSED** [review]

**Review by omridan159 — DISMISSED** [review]

File: app/core/SDKConnect/BatchRPCManager.ts (hunk 2 above)
Comment (abretonc7s) [reply app/core/SDKConnect/BatchRPCManager.ts]: used unknown

**Review by andreahaku — DISMISSED** [review]
LGTM

**Review by andreahaku — APPROVED** [review]
LGTM

**Pull request merged**